### PR TITLE
Hotfix for ads dashboard for sqlondemand

### PR DESCRIPTION
**Pull request by aleklj**
Fixing the ADS dashboard for SqlOnDemand by changing behavior in SqlToolsService

### CODE REVIEW

**Comment by Charles-Gagnon** [comment]
Please make sure you're adding a description for PRs like this - reviewers need the context of what this is doing and why it's necessary.

File: src/Microsoft.SqlTools.ManagedBatchParser/ReliableConnection/ReliableSqlConnection.cs [hunk 1]
```diff
@@ -123,6 +125,21 @@ private bool IsSqlDwConnection(IDbConnection conn)
             return _isSqlDwDatabase;
         }
 
+        /// <summary>
+        /// Determines if a connection is being made to SQLOnDemand.
+        /// </summary>
+        /// <param name="conn">A connection object.</param>
+        private bool IsSqlOnDemandConnection(IDbConnection conn)
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ManagedBatchParser/ReliableConnection/ReliableSqlConnection.cs]: >IsSqlOnDemandConnection [](start = 21, length = 23)

Now we're running the exact same query twice (once for SQL DW and once for On Demand). It'd be better to just query once and store the engine edition information for both after making that call.

Comment (aleklj) [reply src/Microsoft.SqlTools.ManagedBatchParser/ReliableConnection/ReliableSqlConnection.cs]: I think everything is generalized properly now.

File: src/Microsoft.SqlTools.ManagedBatchParser/ReliableConnection/ReliableSqlConnection.cs [hunk 2]
```diff
@@ -104,23 +105,33 @@ protected override void Dispose(bool disposing)
             }
         }
 
+        private DatabaseEngineEdition GetOrReadCachedEngineEdition(IDbConnection conn)
+        {
+            if (CachedServerInfo.Instance.TryGetEngineEdition(conn, out _engineEdition) == DatabaseEngineEdition.Unknown)
+            {
+                _engineEdition = ReliableConnectionHelper.GetEngineEdition(conn);
+                CachedServerInfo.Instance.AddOrUpdateEngineEdition(conn, _engineEdition); ;
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.ManagedBatchParser/ReliableConnection/ReliableSqlConnection.cs]: > ; [](start = 89, length = 2)

extra ; here

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/Connection/CachedServerInfoTests.cs [hunk 3]
```diff
@@ -26,45 +27,43 @@ public void CacheMatchesNullDbNameToEmptyString()
...
-        [InlineData("myDb", true)]  // is SqlDW instance
-        [InlineData(null, false)]   // is not a SqlDw Instance
-        [InlineData("", false)]   // is not a SqlDw Instance
-        [InlineData("myDb", false)]  // is not SqlDW instance
-        public void AddOrUpdateIsSqlDw(string dbName, bool state)
+        [InlineData(null, DatabaseEngineEdition.SqlDataWarehouse)]  // is SqlDW instance
+        [InlineData("", DatabaseEngineEdition.SqlDataWarehouse)]  // is SqlDW instance
+        [InlineData("myDb", DatabaseEngineEdition.SqlDataWarehouse)]  // is SqlDW instance
+        [InlineData(null, DatabaseEngineEdition.SqlOnDemand)]   // is not a SqlDw Instance
+        [InlineData("", DatabaseEngineEdition.SqlOnDemand)]   // is not a SqlDw Instance
```
Comment (Charles-Gagnon) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/Connection/CachedServerInfoTests.cs]: >SqlDw [](start = 74, length = 5)

SqlOnDemand

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/Connection/CachedServerInfoTests.cs [hunk 4]
```diff
@@ -74,41 +73,45 @@ public void AddOrUpdateIsSqlDw(string dbName, bool state)
...
+            cache.AddOrUpdateCache(testSource, state, CachedServerInfo.CacheVariable.EngineEdition);
 
             // Expect the same returned result
-            Assert.True(cache.TryGetIsSqlDw(testSource, out isSqlDwResult));
-            Assert.Equal(isSqlDwResult, state);
+            Assert.NotEqual(cache.TryGetEngineEdition(testSource, out engineEdition), DatabaseEngineEdition.Unknown);
+            Assert.Equal(engineEdition, state);
+
+            DatabaseEngineEdition newState;
+            if (state == DatabaseEngineEdition.SqlDataWarehouse)
```
Comment (Charles-Gagnon) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/Connection/CachedServerInfoTests.cs]: brackets around statements in if

File: src/Microsoft.SqlTools.ManagedBatchParser/ReliableConnection/ReliableConnectionHelper.cs [hunk 5]
```diff
@@ -421,104 +421,55 @@ public static void SetCommandTimeout(IDbCommand cmd)
...
-                              reader.Read();
-                              int engineEditionId = int.Parse(reader[0].ToString(), CultureInfo.InvariantCulture);
-
-                              result = IsCloudEngineId(engineEditionId);
-                          }
+                    commandText,
+                    readResult: (reader) =>
+                    {
+                        reader.Read();
+                        result = (DatabaseEngineEdition)int.Parse(reader[0].ToString(), CultureInfo.InvariantCulture);
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ManagedBatchParser/ReliableConnection/ReliableConnectionHelper.cs]: we don't need to worry about this throwing?  or that's handled somewhere else?

Comment (aleklj) [reply src/Microsoft.SqlTools.ManagedBatchParser/ReliableConnection/ReliableConnectionHelper.cs]: This script has existed in several places for a while now (i just changed the type we're casting to). We probably don't need to worry about it.

**Comment by Charles-Gagnon** [comment]
@aleklj Are you waiting for something? If you want this in the November release you should merge this ASAP so we can get a build out for updating ADS

**Pull request merged**